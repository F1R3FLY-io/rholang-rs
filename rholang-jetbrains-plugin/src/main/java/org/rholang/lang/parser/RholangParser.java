package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
