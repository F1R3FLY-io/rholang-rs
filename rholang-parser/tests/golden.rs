use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
