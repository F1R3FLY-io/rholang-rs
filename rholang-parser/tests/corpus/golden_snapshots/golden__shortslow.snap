---
source: rholang-parser/tests/golden.rs
expression: result
---
Good(
    [
        AnnProc {
            proc: New {
                decls: [
                    NameDecl {
                        id: Id {
                            name: "loop",
                            pos: SourcePos {
                                line: 1,
                                col: 5,
                            },
                        },
                        uri: None,
                    },
                ],
                proc: AnnProc {
                    proc: Par {
                        left: AnnProc {
                            proc: Contract {
                                name: NameVar(
                                    Id(
                                        Id {
                                            name: "loop",
                                            pos: SourcePos {
                                                line: 2,
                                                col: 12,
                                            },
                                        },
                                    ),
                                ),
                                formals: Names {
                                    names: [
                                        Quote(
                                            AnnProc {
                                                proc: ProcVar(
                                                    Id(
                                                        Id {
                                                            name: "n",
                                                            pos: SourcePos {
                                                                line: 2,
                                                                col: 18,
                                                            },
                                                        },
                                                    ),
                                                ),
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 2,
                                                        col: 18,
                                                    },
                                                    end: SourcePos {
                                                        line: 2,
                                                        col: 19,
                                                    },
                                                },
                                            },
                                        ),
                                    ],
                                    remainder: None,
                                },
                                body: AnnProc {
                                    proc: Match {
                                        expression: AnnProc {
                                            proc: ProcVar(
                                                Id(
                                                    Id {
                                                        name: "n",
                                                        pos: SourcePos {
                                                            line: 3,
                                                            col: 11,
                                                        },
                                                    },
                                                ),
                                            ),
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 3,
                                                    col: 11,
                                                },
                                                end: SourcePos {
                                                    line: 3,
                                                    col: 12,
                                                },
                                            },
                                        },
                                        cases: [
                                            Case {
                                                pattern: AnnProc {
                                                    proc: LongLiteral(
                                                        0,
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 4,
                                                            col: 7,
                                                        },
                                                        end: SourcePos {
                                                            line: 4,
                                                            col: 8,
                                                        },
                                                    },
                                                },
                                                proc: AnnProc {
                                                    proc: Nil,
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 4,
                                                            col: 12,
                                                        },
                                                        end: SourcePos {
                                                            line: 4,
                                                            col: 15,
                                                        },
                                                    },
                                                },
                                            },
                                            Case {
                                                pattern: AnnProc {
                                                    proc: ProcVar(
                                                        Wildcard,
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 5,
                                                            col: 7,
                                                        },
                                                        end: SourcePos {
                                                            line: 5,
                                                            col: 8,
                                                        },
                                                    },
                                                },
                                                proc: AnnProc {
                                                    proc: Send {
                                                        channel: NameVar(
                                                            Id(
                                                                Id {
                                                                    name: "loop",
                                                                    pos: SourcePos {
                                                                        line: 5,
                                                                        col: 12,
                                                                    },
                                                                },
                                                            ),
                                                        ),
                                                        send_type: Single,
                                                        inputs: [
                                                            AnnProc {
                                                                proc: BinaryExp {
                                                                    op: Sub,
                                                                    left: AnnProc {
                                                                        proc: ProcVar(
                                                                            Id(
                                                                                Id {
                                                                                    name: "n",
                                                                                    pos: SourcePos {
                                                                                        line: 5,
                                                                                        col: 18,
                                                                                    },
                                                                                },
                                                                            ),
                                                                        ),
                                                                        span: SourceSpan {
                                                                            start: SourcePos {
                                                                                line: 5,
                                                                                col: 18,
                                                                            },
                                                                            end: SourcePos {
                                                                                line: 5,
                                                                                col: 19,
                                                                            },
                                                                        },
                                                                    },
                                                                    right: AnnProc {
                                                                        proc: LongLiteral(
                                                                            1,
                                                                        ),
                                                                        span: SourceSpan {
                                                                            start: SourcePos {
                                                                                line: 5,
                                                                                col: 20,
                                                                            },
                                                                            end: SourcePos {
                                                                                line: 5,
                                                                                col: 21,
                                                                            },
                                                                        },
                                                                    },
                                                                },
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 5,
                                                                        col: 18,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 5,
                                                                        col: 21,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 5,
                                                            col: 12,
                                                        },
                                                        end: SourcePos {
                                                            line: 5,
                                                            col: 22,
                                                        },
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 3,
                                            col: 5,
                                        },
                                        end: SourcePos {
                                            line: 6,
                                            col: 6,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 2,
                                    col: 3,
                                },
                                end: SourcePos {
                                    line: 7,
                                    col: 4,
                                },
                            },
                        },
                        right: AnnProc {
                            proc: Send {
                                channel: NameVar(
                                    Id(
                                        Id {
                                            name: "loop",
                                            pos: SourcePos {
                                                line: 8,
                                                col: 3,
                                            },
                                        },
                                    ),
                                ),
                                send_type: Single,
                                inputs: [
                                    AnnProc {
                                        proc: LongLiteral(
                                            32768,
                                        ),
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 8,
                                                col: 9,
                                            },
                                            end: SourcePos {
                                                line: 8,
                                                col: 14,
                                            },
                                        },
                                    },
                                ],
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 8,
                                    col: 3,
                                },
                                end: SourcePos {
                                    line: 8,
                                    col: 15,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 2,
                            col: 3,
                        },
                        end: SourcePos {
                            line: 8,
                            col: 15,
                        },
                    },
                },
            },
            span: SourceSpan {
                start: SourcePos {
                    line: 1,
                    col: 1,
                },
                end: SourcePos {
                    line: 9,
                    col: 2,
                },
            },
        },
    ],
)
